src/components/application_manager/src/commands/button_notification_to_mobile.cc

```diff
@@ -0,0 +1,209 @@
+  // app_id is mandatory for CUSTOM_BUTTON notification
+  if (!DoesAppIDExist()) {
+    SDL_LOG_ERROR("CUSTOM_BUTTON mobile notification without app_id.");
+    return;
+  }
+
+  // custom_button_id is mandatory for CUSTOM_BUTTON notification
+  if (!(*message_)[msg_params].keyExists(hmi_response::custom_button_id)) {
+    SDL_LOG_ERROR(
+        "CUSTOM_BUTTON mobile notification without custom_button_id.");
+    return;
+  }
+
+  if (!app) {
+    SDL_LOG_ERROR("Application doesn't exist.");
+    return;
+  }
+
+  const uint32_t custom_btn_id =
+      (*message_)[msg_params][hmi_response::custom_button_id].asUInt();
+
+  if (!app->IsSubscribedToSoftButton(custom_btn_id)) {
+    SDL_LOG_ERROR("Application doesn't subscribed to this custom_button_id.");
+    return;
+  }
+
+  if (mobile_api::HMILevel::HMI_NONE ==
+      app->hmi_level(app->GetSoftButtonWindowID(custom_btn_id))) {
+    SDL_LOG_WARN("CUSTOM_BUTTON mobile notification is allowed only "
+                 << "in FULL, LIMITED or BACKGROUND hmi level");
+    return;
+  }
+
+  SendButtonNotification(app);
+  return;
+}
+
+void ButtonNotificationToMobile::HandleOKButton(
+    app_mngr::ApplicationSharedPtr app) {
+  SDL_LOG_AUTO_TRACE();
+  using namespace application_manager;
+  SDL_LOG_DEBUG("OK button received");
+
+  const auto subscribed_apps = SubscribedApps();
+  auto app_ptr = subscribed_apps.end();
+  if (app) {
+    app_ptr = std::find_if(subscribed_apps.begin(),
+                           subscribed_apps.end(),
+                           [&app](const ApplicationSharedPtr subscribed_app) {
+                             return app->app_id() == subscribed_app->app_id();
+                           });
+    if (app_ptr != subscribed_apps.end()) {
+      SDL_LOG_DEBUG(
+          "Sending button press for this app id: " << (*app_ptr)->app_id());
+      SendButtonNotification(*app_ptr);
+      return;
+    }
+  }
+
+  app_ptr = std::find_if(subscribed_apps.begin(),
+                         subscribed_apps.end(),
+                         [](const ApplicationSharedPtr subscribed_app) {
+                           return subscribed_app->IsFullscreen();
+                         });
+
+  if (app_ptr != subscribed_apps.end()) {
+    SDL_LOG_DEBUG("Sending button press for app in FULL with app id: "
+                  << (*app_ptr)->app_id());
+    SendButtonNotification(*app_ptr);
+    return;
+  }
+}
+
+const std::vector<ApplicationSharedPtr>
+ButtonNotificationToMobile::SubscribedApps() const {
+  const uint32_t btn_id = static_cast<uint32_t>(
+      (*message_)[strings::msg_params][hmi_response::button_name].asInt());
+  return application_manager_.applications_by_button(btn_id);
+}
+
+void ButtonNotificationToMobile::HandleMediaButton(
+    app_mngr::ApplicationSharedPtr app) {
+  SDL_LOG_AUTO_TRACE();
+  using namespace application_manager;
+
+  const auto subscribed_apps = SubscribedApps();
+  auto app_ptr = subscribed_apps.end();
```

Comment:
same as above

Reply:
changed in a65c63788fef72c92bd702270a97f27a8f34d4f4